proposals/user-scripts-api.md

```diff
@@ -0,0 +1,189 @@
+  - Present the user with different permission warnings, UI, or other gating, if they deem it necessary.
+
+#### D. Communication between JavaScript worlds
+
+As mentioned in requirement A, the user script world can communicate with different JS worlds via [`window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) and DOM APIs.  New communication methods are being considered as potential future enhancements. This is clunky and imperfect, but allows extensions to migrate to MV3 leveraging the same logic they have in MV2 today.  New, dedicated communication methods are being considered as potential future enhancements.
+
+###  Other considerations
+
+- When multiple scripts match and have the same runAt schedule, the execution order is:
+  - Scripts registered via the content_scripts key in the manifest file
+  - Scripts registered via [`scripting.registerContentScripts()`](https://developer.chrome.com/docs/extensions/reference/scripting/#method-registerContentScripts), following the order they were registered in. Updating a content script doesn't change its registration order.
+  - Scripts registered via `userScripts.register()`, following the order they were registered in. Updating a user script doesn’t change its registration order.
+- User scripts are always persisted across sessions, since the opposite behavior would be uncommon. (We may explore providing an option to customize this in the future.)
+
+### Browser level restrictions
+
+From here, each browser vendor should be able to implement their own restrictions.  Chrome is exploring limiting the access to this API when the user has enabled developer mode (bug), but permission grants are outside of the scope of this API proposal.
+
+## (Potential) Future Enhancements
+
+### `USER_SCRIPT`/ `ISOLATED` World Communication
+
+In the future, we may want to provide a more straightforward path for communication with a `USER_SCRIPT` world (as opposed to the [`window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) approach highlighted above).
+
+### Separate `USER_SCRIPT` Worlds
+
+In addition to specifying the execution world of `USER_SCRIPT`, we could allow extensions to inject in unique worlds by providing an identifier.  Scripts injected with the same identifier would inject in the same world, while scripts with different world identifiers inject in different worlds.  This would allow for greater isolation between user scripts (if, for instance, the user had multiple unrelated user scripts injecting on the same page).
+
+### Execute user scripts one time
+
+Currently, user scripts are registered and executed every time it matches the origin in a persistent way.  We may explore a way to execute a user script only one time to provide a new capability to user scripts (e.g `browser.userScripts.execute()`).
```

Comment:
This is something that is already possible in MV2 (using content scripts and eval). Since the goal of this proposal is to achieve functional parity with MV2, shouldn't this be part of the main proposal?

Reply:
Agreed, this is a feature gap.  We'll open a new PR to discuss it, since a) it may or may not go on the user script API (and might instead go on scripting) and b) it doesn't block development of the User Script API